rq/job.py

```diff
@@ -1308,6 +1326,39 @@ def finished_job_registry(self):
             self.origin, connection=self.connection, job_class=self.__class__, serializer=self.serializer
         )
 
+    def execute_success_callback(self, death_penalty_class: Type[BaseDeathPenalty], result: Any):
+        """Executes success_callback for a job.
+        with timeout .
+
+        Args:
+            death_penalty_class (Type[BaseDeathPenalty]): The penalty class to use for timeout
+            result (Any): The job's result.
+        """
+        if not self.success_callback:
+            return
+
+        logger.debug('Running success callbacks for %s', self.id)
+        self.heartbeat(utcnow(), self.success_callback_timeout)
+        with death_penalty_class(self.success_callback_timeout, JobTimeoutException, job_id=self.id):
+            self.success_callback(self, self.connection, result)
+
+    def execute_failure_callback(self, death_penalty_class: Type[BaseDeathPenalty], *exc_info, heartbeat=False):
+        """Executes failure_callback with possible timeout
+        """
+        if not self.failure_callback:
+            return
+
+        logger.debug('Running failure callbacks for %s', self.id)
+        try:
+            if heartbeat:
+                self.heartbeat(utcnow(), self.failure_callback_timeout)
```

Comment:
Mind moving `heartbeat` out of this method so it's explicitly called by the worker?

Reply:
This requires the timeout to never be None because the worker now uses `job.X_timeout` when sending heartbeat, even if there is no callback defined.
I can add an `if` but if feels weird the worker should know that.